File: Exercicio9/Exercicio9.py
import threading

def soma_intervalo(inicio, fim, resultados, thread_id):
    soma_local = 0
    for j in range(inicio, fim + 1):
        soma_local += j
    resultados[thread_id] = soma_local

num_threads = 4
tamanho_range_threads = 1000000 // num_threads

threads = []
resultados_parciais = [0] * num_threads

for i in range(num_threads):
    inicio = (i * tamanho_range_threads) + 1
    fim = 1000000 if i == num_threads - 1 else (i + 1) * tamanho_range_threads
    t = threading.Thread(target=soma_intervalo, args=(inicio, fim, resultados_parciais, i))
    threads.append(t)
    t.start()

for t in threads:
    t.join()

soma_total = sum(resultados_parciais)
print("Soma total =", soma_total)

File: Exercicio9/Exercicio9.cpp
#include <iostream>
#include <thread>
#include <vector>

using namespace std;

void somaIntervalo(long inicio, long fim, long &resultado) {
    long somaLocal = 0;
    for (long j = inicio; j <= fim; j++) {
        somaLocal += j;
    }
    resultado = somaLocal;
}

int main() {
    const int numThreads = 4;
    long tamanhoRangeThreads = 1000000 / numThreads;

    vector<thread> threads;
    vector<long> resultadosParciais(numThreads);

    for (int i = 0; i < numThreads; i++) {
        long inicio = (i * tamanhoRangeThreads) + 1;
        long fim = (i == numThreads - 1) ? 1000000 : (i + 1) * tamanhoRangeThreads;

        threads.emplace_back(somaIntervalo, inicio, fim, ref(resultadosParciais[i]));
    }

    for (auto &t : threads) {
        t.join();
    }

    long somaTotal = 0;
    for (long r : resultadosParciais) {
        somaTotal += r;
    }

    cout << "Soma total = " << somaTotal << endl;
    return 0;
}
